doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py

```diff
@@ -0,0 +1,569 @@
+                session.report(
+                    {
+                        "epoch": epoch,
+                        "step": step,
+                        "train_loss": loss.item(),
+                        "eval_loss": None,
+                        "perplexity": None,
+                        "number of iterations": step + 1,
+                        "Train time per epoch": None,
+                        "Eval time per epoch": None,
+                        "avg fwd time": avg_fwd_time / (step + 1),
+                        "avg bwd time": avg_bwd_time / (step + 1),
+                    },
+                )
+
+        e_epoch = time.time()
+        accelerator.print("Train time per epoch: ", e_epoch - s_epoch)
+
+        eval_s_epoch = time.time()        
+        valid_dataloader = valid_ds.iter_torch_batches(
+            batch_size=batch_size,
+            collate_fn=collate_partial,
+        )
+        
+        print("Running evaluation ...")
+        perplex, eloss = evaluate(
+            model, 
+            valid_ds_len, 
+            valid_dataloader, 
+            accelerator, 
+            bsize=config["eval_batch_size"],
+            as_test=config["as_test"]
+        )
+        accelerator.print("Eval result loss", eloss)
+        accelerator.print("Eval perplex", perplex)
+
+        eval_e_epoch = time.time()
+        accelerator.print("Eval time per epoch: ", eval_e_epoch - eval_s_epoch)
+        accelerator.print("avg fwd time: ", avg_fwd_time / (step + 1))
+        accelerator.print("avg bwd time: ", avg_bwd_time / (step + 1))
+        accelerator.print("avg opt step time: ", avg_opt_step_time / (step + 1))
+
+        accelerator.print(f"Saving the model in {ckpt_path}")
+        accelerator.wait_for_everyone()
+        with accelerator.main_process_first():
+            ckpt_path_epoch = Path(ckpt_path) / f"epoch-{epoch}"
+            ckpt_path_epoch.mkdir(parents=True, exist_ok=True)
+        
+        if accelerator.is_main_process:
+            print("Saving tokenizer and config.")
+            tokenizer.save_pretrained(ckpt_path_epoch)
+
+        accelerator.wait_for_everyone()
+        
+        # This checkpointing method makes deepspeed checkpoints on each node and then 
+        # Ray Train will aggregate them to a central s3 bucket.
+        # It should be done on all processes (not just the Rank 0)
+        # checkpoint_model(
+        #     checkpoint_folder=ckpt_path_epoch, 
+        #     ckpt_id=epoch, 
+        #     model=model, 
+        #     epoch=epoch, 
+        #     last_global_step=step
+        # )
+        
+        unwrapped_model = accelerator.unwrap_model(model)
+        unwrapped_model.save_pretrained(
+            ckpt_path_epoch,
+            is_main_process=accelerator.is_main_process,
+            save_function=accelerator.save,
+            safe_serialization=True,
+            state_dict=accelerator.get_state_dict(model),
+        )
+        
```

Comment:
I think we do want to checkpoint from each worker rather than just rank 0?

Reply:
Let's merge this to unblock workspace templating ... I need to seat down with you again for the checkpointing.